### DISCUSSION TITLE
Possible to only create MR's when condition is met?

### DISCUSSION THREAD

**Discussion by syphernl** [renovatebot/renovate#9078] [Request Help]
We use our own customized Terraform container and have quite a number of Terraform Infra repo's. 
Our Terraform container repo is being monitored by Renovate as well and creates a new MR when Hashicorp updates their Docker container. 
However, we still need to check that manually (currently) and merge it. After which a new container gets built and pushed to our own registry.

In the mean time for each infra repo an MR is opened updating `.terraform-version` and `versions.tf` but the container also needs to be updated. 
During this Renovate run this container is not available since it hasn't been merged (and built) yet.

Is there a way to "hold off" on creating an MR for `.terraform-version` / `versions.tf` until this particular version is also available in our container registry? 
If the container exists during the second run of Renovate (which is a day later) the existing MR's are properly updated with the container change in `.gitlab-ci.yml`. During it's initial run it results in a failed pipeline, which only gets fixed as when our container is available.

**Comment by rarkins** [comment]
Which Renovate manager is this -`dockerfile`, or `terraform`? I don't fully understand your scenario - it's a little unclear in the above about which parts are updated by Renovate and which bits by your other processes, but let's start with the manager part first. After that, which registry is Renovate *currently* looking up? i.e. the one that shows new versions which aren't available in your internal registry yet.

**Reply by syphernl** [reply]
Thanks for your quick reply!

A config snippet might explain it a little better with how we have it currently setup:

```js
  {
    packageNames: ['hashicorp/terraform', 'registry.gitlab.com/REDACTED/containers/terraform'],
    groupName: 'terraform',
    additionalBranchPrefix: '',
  }
```

This deals with updating `.terraform-version`, `version.tf` and `.gitlab-ci.yml` whenever there is a new Terraform version available and creates a `renovate/terraform` branch with these changes + an MR.

During the run on an infra repo I see the following:
```
 INFO: Dependency extraction complete (repository=REDACTED/infra)
       "baseBranch": "master",
       "stats": {
         "managers": {
           "gitlabci": {"fileCount": 1, "depCount": 2},
           "gitlabci-include": {"fileCount": 1, "depCount": 1},
           "terraform": {"fileCount": 11, "depCount": 35},
           "terraform-version": {"fileCount": 1, "depCount": 1}
         },
         "total": {"fileCount": 14, "depCount": 39}
       }
```

For our Terraform container repo, Renovate updates the Dockerfile with a new `hashicorp/terraform:0.14.X` reference.

**Reply by rarkins** [reply]
Please try to answer the questions I asked carefully. e.g.

> Which Renovate manager is this -`dockerfile`, or `terraform`?

i.e. you are showing me four managers above. Is it one of them, all of them? I frankly can't help if I don't have information on what's being updated. There are a few possibilities but I need to know more before suggesting which to try.

**Reply by syphernl** [reply]
In our Infra repo:
- `.terraform-version` gets updated by the [terraform-version manager](https://docs.renovatebot.com/modules/manager/terraform-version/)
- `version.tf` gets updated by the [terraform manager](https://docs.renovatebot.com/modules/manager/terraform/)
- `.gitlab-ci.yml` gets updated by the [gitlabci manager](https://docs.renovatebot.com/modules/manager/gitlabci/) 

In our Terraform container repo:
- The `Dockerfile` gets updated by the [dockerfile manager](https://docs.renovatebot.com/modules/manager/dockerfile/)

**Reply by rarkins** [reply]
You have 39 dependencies in the log from those four managers. Are all of them affected by this same problem? Or is it just one of them which Renovate thinks is available but it's not available in your private registry?

**Reply by syphernl** [reply]
I think my initial question might have not been clear. Hopefully this example helps with resolving that.

Example:
- Terraform releases `0.14.9`, we currently use `0.14.8` and have this specified in our local files.
- Renovate would normally update `.terraform-version` (terraform-version manager) and `versions.tf` (terraform manager) and replace `0.14.8` with `0.14.9`.
- If we have a new Docker container tag (doesn't really matter which version, as long as its higher than the one in our local files) in our registry, it would also update `.gitlab-ci.yml` (using gitlabci manager)
- This all happens in the same MR because the packageNames group them together.

I would like to have Renovate verify if `container:0.14.9` exists in the registry.
- If the container tag exists: Update the `.gitlab-ci.yml` (alongside the previously changed files). 
  In the end it will update: `.terraform-version`, `versions.tf` and `.gitlab-ci.yml` which all contain a reference to `0.14.9`.
- If the container tag does not exist: Skip creating the MR. Do not update `.terraform-version` or `versions.tf` either.
  During the next Renovate run it will do the check again, if the container tag does exist it would create the MR to update `.terraform-version`, `versions.tf` and `.gitlab-ci.yml`.

**Reply by viceice** [reply]
you need to configure package rule to use your docker image as datasource. maybe you need to disable default manager handling and use custom regex manager to do this.

By default renovate is not able to do that, it will lookup the version on the configured datasources.

**Reply by rarkins** [reply]
Thanks, I understand better now. If Renovate looks up the container using your own internal registry, then it would know there's no 0.14.9. But it doesn't have a way to know that it shouldn't update the others if you use the traditional package managers.

If instead you used regex managers for the other files then you could redirect datasource lookupName to point to your internal *docker* registry so that it doesn't propose an update unless one exists for the Docker image

**Reply by syphernl** [reply]
Would that be as simple as this?

```js
  {
    packageNames: ['hashicorp/terraform', 'registry.gitlab.com/REDACTED/containers/terraform'],
    lookupName: 'registry.gitlab.com/REDACTED/containers/terraform',
    groupName: 'terraform',
    additionalBranchPrefix: '',
  }
```

**Reply by rarkins** [reply]
No, that type of syntax won't work. Please check the documentation and examples for `regexManagers`. You would need to disable this package in its regular managers and extract it using regex manager instead
